python/cugraph/cugraph/sampling/node2vec.py

```diff
@@ -0,0 +1,140 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pylibcugraph
+import cudf
+from cugraph.utilities import ensure_cugraph_obj_for_nx
+
+
+def node2vec(G, start_vertices, max_depth, use_padding, p=1.0, q=1.0):
+    """
+    Computes random walks for each node in 'start_vertices', under the
+    node2vec sampling framework.
+
+    References
+    ----------
+
+    A Grover, J Leskovec: node2vec: Scalable Feature Learning for Networks,
+    Proceedings of the 22nd ACM SIGKDD International Conference on Knowledge
+    Discovery and Data Mining, https://arxiv.org/abs/1607.00653
+
+    Parameters
+    ----------
+    G : cuGraph.Graph or networkx.Graph
+        The graph can be either directed (DiGraph) or undirected (Graph).
+
+    start_vertices: int or list or cudf.Series
+        A single node or a list or a cudf.Series of nodes from which to run
+        the random walks
+
+    max_depth: int, optional
+        The maximum depth of the random walks
+
+    use_padding: bool, optional
+        If True, padded paths are returned else coalesced paths are returned
+
+    p: double, optional
+        Return factor, which represents the likelihood of backtracking to
+        a previous node in the walk. A higher value makes it less likely to
+        sample a previously visited node, while a lower value makes it more
+        likely to backtrack, making the walk "local"
+
+    q: double, optional
+        In-out factor, which represents the likelihood of visiting nodes
+        closer or further from the outgoing node. If q > 1, the random walk
+        is likelier to visit nodes closer to the outgoing node. If q < 1, the
```

Comment:
if the `Return` and `In-out factor` are required to be positive double, probably mention it in the docsctring like 
`0 <= q < 1` so that the user is aware of it?

Reply:
Because both p and q represent likelihoods and not probabilities, they aren't bound like `0 <= q < 1` but rather `0 < p, q`. I can include it in the argument description

Reply:
I was saying `0 <= q < 1` because I thought it needed to be a positive double?

from your code
```
if (q is None) or (q <= 0.0):
        raise ValueError("'q' must be a positive double")
```

Reply:
Both p and q can be greater than 1, so as long as they aren't None or a negative value, the ValueError wouldn't be raised

Reply:
Got it. I didn't understand your comment at first. That it should just be positive double

Reply:
Yes, it's a positive double in pylibcugraph bc of using C types, but in python it's a float (same precision). Either way, it's been clarified in a later description